rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,337 @@
+in Kibana.
+
+# Basic example
+
+```ts
+// Override default behavior and only elevate severity when elasticsearch is not available
+core.status.set(
+  core.status.core$.pipe(core => core.elasticsearch);
+)
+```
+
+# Motivation
+
+Kibana should do as much possible to help users keep their installation in a working state. This includes providing as much detail about components that are not working as well as ensuring that failures in one part of the application do not block using other portions of the application.
+
+In order to provide the user with as much detail as possible about any systems that are not working correctly, the status mechanism should provide excellent defaults in terms of expressing relationships between services and presenting detailed information to the user.
+
+# Detailed design
+
+## Types
+
+```ts
+/**
+ * The current status of a service at a point in time.
+ */
+type ServiceStatus = {
+  /**
+   * The current availability level of the service.
+   */
+  level: ServiceStatusLevel.available;
+  /**
+   * A high-level summary of the service status.
+   */
+  summary?: string;
+  /**
+   * A more detailed description of the service status.
+   */
+  detail?: string;
+  /**
+   * A URL to open in a new tab about how to resolve or troubleshoot the problem.
+   */
+  documentationUrl?: string;
+  /**
+   * Any JSON-serializable data to be included in the HTTP API response. Useful for providing more fine-grained, 
+   * machine-readable information about the service status. May include status information for underlying features.
+   */
+  meta?: JSONObject;
+} | {
+  level: ServiceStatusLevel;
+  summary: string; // required when level !== available
+  detail?: string;
+  documentationUrl?: string;
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  /**
+   * Everything is working!
+   */
+  available,
+  /**
+   * Some features may not be working.
+   */
+  degraded,
+  /**
+   * The service is unavailable, but other functions that do not depend on this service should work.
+   */
+  unavailable,
+  /**
+   * Block all user functions and display the status page, reserved for Core services only.
+   * Note: In the real implementation, this will be split out to a different type. Kept as a single type here to make
+   * the RFC easier to follow.
+   */
+  critical
+}
+
+/**
+ * Status of core services. Only contains entries for backend services that could have a non-available `status`.
+ * For example, `context` cannot possibly be broken, so it is not included.
+ */
+interface CoreStatus {
+  elasticsearch: ServiceStatus;
+  http: ServiceStatus;
+  savedObjects: ServiceStatus;
+  uiSettings: ServiceStatus;
+  metrics: ServiceStatus;
+}
+
+// Types that specify valid JSON values.
+type JSONValue = string | number | boolean | null | JSONValue[] | JSONObject;
+type JSONObject = Record<string, JSONValue>;
```

Comment:
Ugh I forgot about that problem in recent versions of TS. 

> I think it would be more useful to type this with a generic to make it easier for other plugins to consume this meta information.

So each plugin could export a type that describes the shape of its `meta` field which other plugins could use. Makes sense to me 👍